Plant (replacement for MiGardener)

ChristianKuehnel opened the pull request:
## Description:
New implementation of plant monitoring component, as requested in comment here:
https://github.com/home-assistant/home-assistant/pull/6684#issuecomment-289987669

## Example entry for `configuration.yaml` (if applicable):
```yaml
plant:
  simulated_plant:
    sensors:
      moisture: sensor.mqtt_plant_moisture
      battery: sensor.mqtt_plant_battery
      temperature: sensor.mqtt_plant_temperature
      conductivity: sensor.mqtt_plant_conductivity
      brightness: sensor.mqtt_plant_brightness
    min_moisture: 20
    max_moisture: 60
    min_battery: 17
    min_conductivity: 500
    min_temperature: 15
```
#### Using plain MQTT sensor to get the data

```yaml
sensor:
  - platform: mqtt
    name: mqtt_plant_moisture
    state_topic: test/simulated_plant
    value_template: '{{ value_json.moisture }}'
  - platform: mqtt
    name: mqtt_plant_battery
    state_topic: test/simulated_plant
    value_template: '{{ value_json.battery }}'
  - platform: mqtt
    name: mqtt_plant_temperature
    state_topic: test/simulated_plant
    value_template: '{{ value_json.temperature }}'
  - platform: mqtt
    name: mqtt_plant_conductivity
    state_topic: test/simulated_plant
    value_template: '{{ value_json.conductivity }}'
  - platform: mqtt
    name: mqtt_plant_brightness
    state_topic: test/simulated_plant
    value_template: '{{ value_json.brightness }}'
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

ChristianKuehnel committed:
new implementation without mqtt

ChristianKuehnel committed:
fixed lint findings

ChristianKuehnel committed:
fixed more lint findings

ChristianKuehnel committed:
fixed final flak8 error

ChristianKuehnel committed:
added unit tests for platform "plant"

fabaff commented:
I don't know what was discussed in #6684 but i guess that the code below could help if somebody want to simulate `plantgateway` without setting it up.

```python
from datetime import datetime
import json
import random

import paho.mqtt.client as mqtt

client = mqtt.Client()
client.connect('localhost', 1883, 60)

send_msg = {
    'battery': random.randrange(80, 100, 1),
    'temperature': random.randrange(0, 30, 1),
    'brightness': random.randrange(40, 100, 1),
    'moisture': random.randrange(10, 70, 1),
    'conductivity': random.randrange(40, 70, 1),
    'timestamp': datetime.now().isoformat(),
}

client.publish('test/simulated_plant', payload=json.dumps(send_msg), qos=0, retain=False)
```

fabaff commented:
For me this component looks like a triary sensor as the state is ok or high/low. At the moment I will not work well with the frontend.

![screenshot from 2017-04-23 21-46-03](https://cloud.githubusercontent.com/assets/116184/25316899/018e5a90-286f-11e7-9a9c-c5bca0f7bfa0.png)

Wouldn't it be an option to just show that there is deviation (ok/not ok) like a binary sensor? And then show the state which is too low/high as attribute. 

Sorry, I will read  #6684 before I add more comments.

ChristianKuehnel commented:
@fabaff 

Thank you for the feedback.

The component will output a string that says "ok" if all readings are within the intended range or "xxx high/low" if they are outside of the range. As there might be several problems in paralle, you will get "battery low, moisture low, brightness high".

I like youre idea about setting state to "ok" and "problem". Then I can provide more details in the attributes (e.g. for sending notifications). 

This is also what I'm using it for right now: if state =! ok, send notification with the details of the problem

balloob commented:
The code looks great ! 🐬 

This can be merged once the state has been updated to be `ok` / `problem`

ChristianKuehnel committed:
- changed status to "OK" / "problem"
- added attribute "problem" with details on the problems
- removed unused constant
- setting icon to "?" until we have meaningful data

ChristianKuehnel committed:
reformatted code to meet line length requirements

ChristianKuehnel commented on tests/components/test_plant.py:
```diff
@@ -0,0 +1,72 @@
+                                  TestPlant._MockState(45),
+                                  TestPlant._MockState(10))
+        self.assertEqual(self.sensor.state, 'problem')
+        self.assertEqual(self.sensor.state_attributes['problem'], 'battery low')
```
fixed

ChristianKuehnel commented:
@balloob thx for the review, I implemented the changes as requested

balloob commented:
Awesome 🐬

balloob commented:
No documentation was ever submitted for this feature 😢

ChristianKuehnel commented:
created pull request for documentation: https://github.com/home-assistant/home-assistant.github.io/pull/2595
